Linting code with Clippy does not properly account for the selected Rust edition

# What I did

Ran Clippy on the following Rust 2018 code:

```rust
use base64::STANDARD; // 0.10.0

fn main() {
    println!("{:?}", STANDARD);
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3d9ce7f5687f2e278e66c6cfe837b5f3))

# What happened

Clippy complains about a missing external crate:

```
error[E0432]: unresolved import `base64`
 --> src/main.rs:1:5
  |
1 | use base64::STANDARD; // 0.10.0
  |     ^^^^^^ maybe a missing `extern crate base64;`?
``` 

# What I expected to happen

I expected the code to be validly parsed and Clippy warnings to be printed (if any).

# Additional information

The  code compiles/runs fine when using "Run".

--- a/tests/Gemfile
+++ b/tests/Gemfile
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 source "https://rubygems.org"
-ruby "2.5.1"
+ruby "2.6.0"
 
 gem "capybara"
 gem "rspec"

--- a/tests/spec/features/tools_spec.rb
+++ b/tests/spec/features/tools_spec.rb
@@ -30,7 +30,9 @@ RSpec.feature "Using third-party Rust tools", type: :feature, js: true do
 
   def code_with_lint_warnings
     <<~EOF
-    fn main() {
+    use itertools::Itertools;
+
+    fn example() {
         let a = 0.0 / 0.0;
         println!("NaN is {}", a);
     }

--- a/tests/spec/features/tools_spec.rb
+++ b/tests/spec/features/tools_spec.rb
@@ -42,7 +42,7 @@ RSpec.feature "Using third-party Rust tools", type: :feature, js: true do
     in_tools_menu { click_on("Miri") }
 
     within(".output-stderr") do
-      expect(page).to have_content /pointer computed at offset 1, outside bounds of allocation \d+ which has size 0/
+      expect(page).to have_content %r{Pointer must be in-bounds and live at offset 1, but is outside bounds of allocation \d+ which has size 0}
     end
   end
 

--- a/tests/spec/features/url_parameters_spec.rb
+++ b/tests/spec/features/url_parameters_spec.rb
@@ -59,14 +59,14 @@ RSpec.feature "Configuration by URL parameters", type: :feature, js: true do
   end
 
   scenario "loading from a Gist with an edition preserves the edition" do
-    visit '/?gist=20fb1e0475f890d0fdb7864e3ad0820c&version=nightly&edition=2018'
+    visit '/?gist=20fb1e0475f890d0fdb7864e3ad0820c&edition=2018'
 
     expect(page).to have_edition('2018')
     expect(page).to have_link("Permalink to the playground", href: /edition=2018/)
   end
 
   scenario "loading from a Gist without an edition selects Rust 2015" do
-    visit '/?gist=20fb1e0475f890d0fdb7864e3ad0820c&version=nightly'
+    visit '/?gist=20fb1e0475f890d0fdb7864e3ad0820c'
 
     expect(page).to have_edition('2015')
     expect(page).to have_link("Permalink to the playground", href: /edition=2015/)
@@ -79,7 +79,7 @@ RSpec.feature "Configuration by URL parameters", type: :feature, js: true do
   end
 
   scenario "loading code directly from a parameter without an edition selects Rust 2015" do
-    visit '/?code=fn%20main()%20%7B%0A%20%20%20%20println!(%22Hello%2C%20world!%22)%3B%0A%7D&version=nightly'
+    visit '/?code=fn%20main()%20%7B%0A%20%20%20%20println!(%22Hello%2C%20world!%22)%3B%0A%7D'
 
     expect(page).to have_edition('2015')
   end
@@ -97,23 +97,14 @@ RSpec.feature "Configuration by URL parameters", type: :feature, js: true do
   end
 
   scenario "loading with an edition" do
-    visit '/?version=nightly&edition=2018'
+    visit '/?edition=2018'
 
     expect(page).to have_edition('2018')
   end
 
-  scenario "loading with an edition on non-nightly" do
-    # Once stable and beta Cargos understand the edition marker,
-    # remove this test, the `version=nightly` from the other edition
-    # tests, and the related code.
-    visit '/?edition=2018'
-
-    expect(page).to have_edition('2015')
-  end
-
   scenario "loading without code or an edition" do
     visit '/'
-    expect(page).to have_edition('2015')
+    expect(page).to have_edition('2018')
   end
 
   def editor

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -222,7 +222,7 @@ interface ExecuteRequestBody {
   crateType: string;
   tests: boolean;
   code: string;
-  edition?: string;
+  edition: string;
   backtrace: boolean;
 }
 
@@ -481,7 +481,7 @@ const requestMiri = () =>
 
 interface MiriRequestBody {
   code: string;
-  edition?: string;
+  edition: string;
 }
 
 const receiveMiriSuccess = ({ stdout, stderr }) =>

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -2,7 +2,7 @@ import fetch from 'isomorphic-fetch';
 import { ThunkAction as ReduxThunkAction } from 'redux-thunk';
 import url from 'url';
 
-import { getCrateType, isAutoBuildSelector, runAsTest } from './selectors';
+import { clippyRequestSelector, getCrateType, isAutoBuildSelector, runAsTest } from './selectors';
 import State from './state';
 import {
   AssemblyFlavor,
@@ -456,6 +456,12 @@ export function performFormat(): ThunkAction {
 const requestClippy = () =>
   createAction(ActionType.RequestClippy);
 
+interface ClippyRequestBody {
+  code: string;
+  edition: string;
+  crateType: string;
+}
+
 const receiveClippySuccess = ({ stdout, stderr }) =>
   createAction(ActionType.ClippySucceeded, { stdout, stderr });
 
@@ -467,8 +473,7 @@ export function performClippy(): ThunkAction {
   return function(dispatch, getState) {
     dispatch(requestClippy());
 
-    const { code } = getState();
-    const body = { code };
+    const body: ClippyRequestBody = clippyRequestSelector(getState());
 
     return jsonPost(routes.clippy, body)
       .then(json => dispatch(receiveClippySuccess(json)))

--- a/ui/frontend/selectors/index.ts
+++ b/ui/frontend/selectors/index.ts
@@ -100,6 +100,8 @@ export const clippyVersionDetailsText = createSelector([getClippy], versionDetai
 export const rustfmtVersionDetailsText = createSelector([getRustfmt], versionDetails);
 export const miriVersionDetailsText = createSelector([getMiri], versionDetails);
 
+const editionSelector = (state: State) => state.configuration.edition;
+
 export const isWasmAvailable = (state: State) => (
   state.configuration.channel === Channel.Nightly
 );
@@ -114,8 +116,9 @@ export const getChannelLabel = (state: State) => {
   return `${channel}`;
 };
 
-export const getEditionSet = (state: State) => (
-  state.configuration.edition !== Edition.Rust2018
+export const getEditionSet = createSelector(
+  editionSelector,
+  edition => edition !== Edition.Rust2018,
 );
 
 export const getBacktraceSet = (state: State) => (
@@ -260,4 +263,11 @@ export const showRust2018IsDefaultSelector = createSelector(
 export const anyNotificationsToShowSelector = createSelector(
   showRust2018IsDefaultSelector,
   allNotifications => allNotifications,
+);
+
+export const clippyRequestSelector = createSelector(
+  codeSelector,
+  editionSelector,
+  getCrateType,
+  (code, edition, crateType) => ({ code, edition, crateType }),
 );

--- a/ui/src/main.rs
+++ b/ui/src/main.rs
@@ -192,7 +192,7 @@ fn format(req: &mut Request) -> IronResult<Response> {
 fn clippy(req: &mut Request) -> IronResult<Response> {
     with_sandbox(req, |sandbox, req: ClippyRequest| {
         sandbox
-            .clippy(&req.into())
+            .clippy(&req.try_into()?)
             .map(ClippyResponse::from)
             .map_err(Error::Sandbox)
     })
@@ -633,6 +633,10 @@ struct FormatResponse {
 #[derive(Debug, Clone, Deserialize)]
 struct ClippyRequest {
     code: String,
+    #[serde(default)]
+    edition: String,
+    #[serde(default = "default_crate_type", rename = "crateType")]
+    crate_type: String,
 }
 
 #[derive(Debug, Clone, Serialize)]
@@ -797,11 +801,15 @@ impl From<sandbox::FormatResponse> for FormatResponse {
     }
 }
 
-impl From<ClippyRequest> for sandbox::ClippyRequest {
-    fn from(me: ClippyRequest) -> Self {
-        sandbox::ClippyRequest {
+impl TryFrom<ClippyRequest> for sandbox::ClippyRequest {
+    type Error = Error;
+
+    fn try_from(me: ClippyRequest) -> Result<Self> {
+        Ok(sandbox::ClippyRequest {
             code: me.code,
-        }
+            crate_type: parse_crate_type(&me.crate_type)?,
+            edition: parse_edition(&me.edition)?,
+        })
     }
 }
 
@@ -984,4 +992,8 @@ fn parse_crate_type(s: &str) -> Result<sandbox::CrateType> {
         "proc-macro" => Library(ProcMacro),
         _ => return Err(Error::InvalidCrateType(s.into()))
     })
+}
+
+fn default_crate_type() -> String {
+    "bin".into()
 }

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -120,7 +120,7 @@ impl Sandbox {
     pub fn compile(&self, req: &CompileRequest) -> Result<CompileResponse> {
         try!(self.write_source_code(&req.code));
 
-        let mut command = self.compile_command(req.target, req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req);
+        let mut command = self.compile_command(req.target, req.channel, req.mode, req.tests, req.backtrace, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -172,7 +172,7 @@ impl Sandbox {
 
     pub fn execute(&self, req: &ExecuteRequest) -> Result<ExecuteResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req);
+        let mut command = self.execute_command(req.channel, req.mode, req.tests, req.backtrace, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -306,11 +306,11 @@ impl Sandbox {
         Ok(())
     }
 
-    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, req: impl EditionRequest) -> Command {
-        let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, Some(target), crate_type, req, backtrace);
+    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, tests: bool, backtrace: bool, req: impl CrateTypeRequest + EditionRequest) -> Command {
+        let mut cmd = self.docker_command(Some(req.crate_type()));
+        set_execution_environment(&mut cmd, Some(target), &req, backtrace);
 
-        let execution_cmd = build_execution_command(Some(target), channel, mode, crate_type, tests);
+        let execution_cmd = build_execution_command(Some(target), channel, mode, &req, tests);
 
         cmd.arg(&channel.container_name()).args(&execution_cmd);
 
@@ -319,11 +319,11 @@ impl Sandbox {
         cmd
     }
 
-    fn execute_command(&self, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, req: impl EditionRequest) -> Command {
-        let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, None, crate_type, req, backtrace);
+    fn execute_command(&self, channel: Channel, mode: Mode, tests: bool, backtrace: bool, req: impl CrateTypeRequest + EditionRequest) -> Command {
+        let mut cmd = self.docker_command(Some(req.crate_type()));
+        set_execution_environment(&mut cmd, None, &req, backtrace);
 
-        let execution_cmd = build_execution_command(None, channel, mode, crate_type, tests);
+        let execution_cmd = build_execution_command(None, channel, mode, &req, tests);
 
         cmd.arg(&channel.container_name()).args(&execution_cmd);
 
@@ -409,14 +409,14 @@ fn basic_secure_docker_command() -> Command {
     cmd
 }
 
-fn build_execution_command(target: Option<CompileTarget>, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool) -> Vec<&'static str> {
+fn build_execution_command(target: Option<CompileTarget>, channel: Channel, mode: Mode, req: impl CrateTypeRequest, tests: bool) -> Vec<&'static str> {
     use self::CompileTarget::*;
     use self::CrateType::*;
     use self::Mode::*;
 
     let mut cmd = vec!["cargo"];
 
-    match (target, crate_type, tests) {
+    match (target, req.crate_type(), tests) {
         (Some(Wasm), _, _) => cmd.push("wasm"),
         (Some(_), _, _)    => cmd.push("rustc"),
         (_, _, true)       => cmd.push("test"),
@@ -458,20 +458,16 @@ fn build_execution_command(target: Option<CompileTarget>, channel: Channel, mode
     cmd
 }
 
-fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, crate_type: CrateType, req: impl EditionRequest, backtrace: bool) {
+fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, req: impl CrateTypeRequest + EditionRequest, backtrace: bool) {
     use self::CompileTarget::*;
-    use self::CrateType::*;
 
     if let Some(Wasm) = target {
         cmd.args(&["--env", "PLAYGROUND_NO_DEPENDENCIES=true"]);
         cmd.args(&["--env", "PLAYGROUND_RELEASE_LTO=true"]);
     }
 
-    if let Library(lib) = crate_type {
-        cmd.args(&["--env", &format!("PLAYGROUND_CRATE_TYPE={}", lib.cargo_ident())]);
-    }
-
-    cmd.apply_edition(req);
+    cmd.apply_crate_type(&req);
+    cmd.apply_edition(&req);
 
     if backtrace {
         cmd.args(&["--env", "RUST_BACKTRACE=1"]);
@@ -626,15 +622,30 @@ impl LibraryType {
 }
 
 trait DockerCommandExt {
+    fn apply_crate_type(&mut self, req: impl CrateTypeRequest);
     fn apply_edition(&mut self, req: impl EditionRequest);
 }
 
 impl DockerCommandExt for Command {
+    fn apply_crate_type(&mut self, req: impl CrateTypeRequest) {
+        if let CrateType::Library(lib) = req.crate_type() {
+            self.args(&["--env", &format!("PLAYGROUND_CRATE_TYPE={}", lib.cargo_ident())]);
+        }
+    }
+
     fn apply_edition(&mut self, req: impl EditionRequest) {
         if let Some(edition) = req.edition() {
             self.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
         }
     }
+}
+
+trait CrateTypeRequest {
+    fn crate_type(&self) -> CrateType;
+}
+
+impl<R: CrateTypeRequest> CrateTypeRequest for &'_ R {
+    fn crate_type(&self) -> CrateType { (*self).crate_type() }
 }
 
 trait EditionRequest {
@@ -655,6 +666,10 @@ pub struct CompileRequest {
     pub tests: bool,
     pub backtrace: bool,
     pub code: String,
+}
+
+impl CrateTypeRequest for CompileRequest {
+    fn crate_type(&self) -> CrateType { self.crate_type }
 }
 
 impl EditionRequest for CompileRequest {
@@ -678,6 +693,10 @@ pub struct ExecuteRequest {
     pub tests: bool,
     pub backtrace: bool,
     pub code: String,
+}
+
+impl CrateTypeRequest for ExecuteRequest {
+    fn crate_type(&self) -> CrateType { self.crate_type }
 }
 
 impl EditionRequest for ExecuteRequest {

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -199,7 +199,7 @@ impl Sandbox {
 
     pub fn clippy(&self, req: &ClippyRequest) -> Result<ClippyResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.clippy_command();
+        let mut command = self.clippy_command(req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -344,8 +344,11 @@ impl Sandbox {
         cmd
     }
 
-    fn clippy_command(&self) -> Command {
-        let mut cmd = self.docker_command(None);
+    fn clippy_command(&self, req: impl CrateTypeRequest + EditionRequest) -> Command {
+        let mut cmd = self.docker_command(Some(req.crate_type()));
+
+        cmd.apply_crate_type(&req);
+        cmd.apply_edition(&req);
 
         cmd.arg("clippy").args(&["cargo", "clippy"]);
 
@@ -746,6 +749,16 @@ pub struct FormatResponse {
 #[derive(Debug, Clone)]
 pub struct ClippyRequest {
     pub code: String,
+    pub edition: Option<Edition>,
+    pub crate_type: CrateType,
+}
+
+impl CrateTypeRequest for ClippyRequest {
+    fn crate_type(&self) -> CrateType { self.crate_type }
+}
+
+impl EditionRequest for ClippyRequest {
+    fn edition(&self) -> Option<Edition> { self.edition }
 }
 
 #[derive(Debug, Clone)]
@@ -807,6 +820,16 @@ mod test {
                 code: HELLO_WORLD_CODE.to_string(),
                 edition: None,
                 backtrace: false,
+            }
+        }
+    }
+
+    impl Default for ClippyRequest {
+        fn default() -> Self {
+            ClippyRequest {
+                code: HELLO_WORLD_CODE.to_string(),
+                crate_type: CrateType::Binary,
+                edition: None,
             }
         }
     }
@@ -1105,6 +1128,31 @@ mod test {
 
         let req = ClippyRequest {
             code: code.to_string(),
+            ..ClippyRequest::default()
+        };
+
+        let sb = Sandbox::new().expect("Unable to create sandbox");
+        let resp = sb.clippy(&req).expect("Unable to lint code");
+
+        assert!(resp.stderr.contains("deny(clippy::eq_op)"));
+        assert!(resp.stderr.contains("warn(clippy::zero_divided_by_zero)"));
+    }
+
+    #[test]
+    fn linting_code_options() {
+        let code = r#"
+        use itertools::Itertools; // Edition 2018 feature
+
+        fn example() {
+            let a = 0.0 / 0.0;
+            println!("NaN is {}", a);
+        }
+        "#;
+
+        let req = ClippyRequest {
+            code: code.to_string(),
+            crate_type: CrateType::Library(LibraryType::Rlib),
+            edition: Some(Edition::Rust2018),
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -120,7 +120,7 @@ impl Sandbox {
     pub fn compile(&self, req: &CompileRequest) -> Result<CompileResponse> {
         try!(self.write_source_code(&req.code));
 
-        let mut command = self.compile_command(req.target, req.channel, req.mode, req.tests, req.backtrace, req);
+        let mut command = self.compile_command(req.target, req.channel, req.mode, req.tests, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -172,7 +172,7 @@ impl Sandbox {
 
     pub fn execute(&self, req: &ExecuteRequest) -> Result<ExecuteResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.execute_command(req.channel, req.mode, req.tests, req.backtrace, req);
+        let mut command = self.execute_command(req.channel, req.mode, req.tests, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -306,9 +306,9 @@ impl Sandbox {
         Ok(())
     }
 
-    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, tests: bool, backtrace: bool, req: impl CrateTypeRequest + EditionRequest) -> Command {
+    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, tests: bool, req: impl CrateTypeRequest + EditionRequest + BacktraceRequest) -> Command {
         let mut cmd = self.docker_command(Some(req.crate_type()));
-        set_execution_environment(&mut cmd, Some(target), &req, backtrace);
+        set_execution_environment(&mut cmd, Some(target), &req);
 
         let execution_cmd = build_execution_command(Some(target), channel, mode, &req, tests);
 
@@ -319,9 +319,9 @@ impl Sandbox {
         cmd
     }
 
-    fn execute_command(&self, channel: Channel, mode: Mode, tests: bool, backtrace: bool, req: impl CrateTypeRequest + EditionRequest) -> Command {
+    fn execute_command(&self, channel: Channel, mode: Mode, tests: bool, req: impl CrateTypeRequest + EditionRequest + BacktraceRequest) -> Command {
         let mut cmd = self.docker_command(Some(req.crate_type()));
-        set_execution_environment(&mut cmd, None, &req, backtrace);
+        set_execution_environment(&mut cmd, None, &req);
 
         let execution_cmd = build_execution_command(None, channel, mode, &req, tests);
 
@@ -458,7 +458,7 @@ fn build_execution_command(target: Option<CompileTarget>, channel: Channel, mode
     cmd
 }
 
-fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, req: impl CrateTypeRequest + EditionRequest, backtrace: bool) {
+fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, req: impl CrateTypeRequest + EditionRequest + BacktraceRequest) {
     use self::CompileTarget::*;
 
     if let Some(Wasm) = target {
@@ -468,10 +468,7 @@ fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, r
 
     cmd.apply_crate_type(&req);
     cmd.apply_edition(&req);
-
-    if backtrace {
-        cmd.args(&["--env", "RUST_BACKTRACE=1"]);
-    }
+    cmd.apply_backtrace(&req);
 }
 
 fn read(path: &Path) -> Result<Option<String>> {
@@ -624,6 +621,7 @@ impl LibraryType {
 trait DockerCommandExt {
     fn apply_crate_type(&mut self, req: impl CrateTypeRequest);
     fn apply_edition(&mut self, req: impl EditionRequest);
+    fn apply_backtrace(&mut self, req: impl BacktraceRequest);
 }
 
 impl DockerCommandExt for Command {
@@ -636,6 +634,12 @@ impl DockerCommandExt for Command {
     fn apply_edition(&mut self, req: impl EditionRequest) {
         if let Some(edition) = req.edition() {
             self.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
+        }
+    }
+
+    fn apply_backtrace(&mut self, req: impl BacktraceRequest) {
+        if req.backtrace() {
+            self.args(&["--env", "RUST_BACKTRACE=1"]);
         }
     }
 }
@@ -656,6 +660,14 @@ impl<R: EditionRequest> EditionRequest for &'_ R {
     fn edition(&self) -> Option<Edition> { (*self).edition() }
 }
 
+trait BacktraceRequest {
+    fn backtrace(&self) -> bool;
+}
+
+impl<R: BacktraceRequest> BacktraceRequest for &'_ R {
+    fn backtrace(&self) -> bool { (*self).backtrace() }
+}
+
 #[derive(Debug, Clone)]
 pub struct CompileRequest {
     pub target: CompileTarget,
@@ -674,6 +686,10 @@ impl CrateTypeRequest for CompileRequest {
 
 impl EditionRequest for CompileRequest {
     fn edition(&self) -> Option<Edition> { self.edition }
+}
+
+impl BacktraceRequest for CompileRequest {
+    fn backtrace(&self) -> bool { self.backtrace }
 }
 
 #[derive(Debug, Clone)]
@@ -701,6 +717,10 @@ impl CrateTypeRequest for ExecuteRequest {
 
 impl EditionRequest for ExecuteRequest {
     fn edition(&self) -> Option<Edition> { self.edition }
+}
+
+impl BacktraceRequest for ExecuteRequest {
+    fn backtrace(&self) -> bool { self.backtrace }
 }
 
 #[derive(Debug, Clone)]

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -120,7 +120,7 @@ impl Sandbox {
     pub fn compile(&self, req: &CompileRequest) -> Result<CompileResponse> {
         try!(self.write_source_code(&req.code));
 
-        let mut command = self.compile_command(req.target, req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
+        let mut command = self.compile_command(req.target, req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -172,7 +172,7 @@ impl Sandbox {
 
     pub fn execute(&self, req: &ExecuteRequest) -> Result<ExecuteResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
+        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -212,7 +212,7 @@ impl Sandbox {
 
     pub fn miri(&self, req: &MiriRequest) -> Result<MiriResponse> {
         self.write_source_code(&req.code)?;
-        let mut command = self.miri_command(req.edition);
+        let mut command = self.miri_command(req);
 
         let output = command.output().map_err(Error::UnableToExecuteCompiler)?;
 
@@ -306,9 +306,9 @@ impl Sandbox {
         Ok(())
     }
 
-    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, edition: Option<Edition>) -> Command {
+    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, req: impl EditionRequest) -> Command {
         let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, Some(target), crate_type, edition, backtrace);
+        set_execution_environment(&mut cmd, Some(target), crate_type, req, backtrace);
 
         let execution_cmd = build_execution_command(Some(target), channel, mode, crate_type, tests);
 
@@ -319,9 +319,9 @@ impl Sandbox {
         cmd
     }
 
-    fn execute_command(&self, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, edition: Option<Edition>) -> Command {
+    fn execute_command(&self, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, req: impl EditionRequest) -> Command {
         let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, None, crate_type, edition, backtrace);
+        set_execution_environment(&mut cmd, None, crate_type, req, backtrace);
 
         let execution_cmd = build_execution_command(None, channel, mode, crate_type, tests);
 
@@ -354,13 +354,9 @@ impl Sandbox {
         cmd
     }
 
-    fn miri_command(&self, edition: Option<Edition>) -> Command {
+    fn miri_command(&self, req: impl EditionRequest) -> Command {
         let mut cmd = self.docker_command(None);
-        if let Some(edition) = edition {
-            // set the env var that will cause modify-cargo-toml inside the docker
-            // container to set this in the `Cargo.toml`.
-            cmd.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
-        }
+        cmd.apply_edition(req);
 
         cmd.arg("miri").args(&["cargo", "miri-playground"]);
 
@@ -462,7 +458,7 @@ fn build_execution_command(target: Option<CompileTarget>, channel: Channel, mode
     cmd
 }
 
-fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, crate_type: CrateType, edition: Option<Edition>, backtrace: bool) {
+fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, crate_type: CrateType, req: impl EditionRequest, backtrace: bool) {
     use self::CompileTarget::*;
     use self::CrateType::*;
 
@@ -475,9 +471,7 @@ fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, c
         cmd.args(&["--env", &format!("PLAYGROUND_CRATE_TYPE={}", lib.cargo_ident())]);
     }
 
-    if let Some(edition) = edition {
-        cmd.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
-    }
+    cmd.apply_edition(req);
 
     if backtrace {
         cmd.args(&["--env", "RUST_BACKTRACE=1"]);
@@ -631,6 +625,26 @@ impl LibraryType {
     }
 }
 
+trait DockerCommandExt {
+    fn apply_edition(&mut self, req: impl EditionRequest);
+}
+
+impl DockerCommandExt for Command {
+    fn apply_edition(&mut self, req: impl EditionRequest) {
+        if let Some(edition) = req.edition() {
+            self.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
+        }
+    }
+}
+
+trait EditionRequest {
+    fn edition(&self) -> Option<Edition>;
+}
+
+impl<R: EditionRequest> EditionRequest for &'_ R {
+    fn edition(&self) -> Option<Edition> { (*self).edition() }
+}
+
 #[derive(Debug, Clone)]
 pub struct CompileRequest {
     pub target: CompileTarget,
@@ -641,6 +655,10 @@ pub struct CompileRequest {
     pub tests: bool,
     pub backtrace: bool,
     pub code: String,
+}
+
+impl EditionRequest for CompileRequest {
+    fn edition(&self) -> Option<Edition> { self.edition }
 }
 
 #[derive(Debug, Clone)]
@@ -660,6 +678,10 @@ pub struct ExecuteRequest {
     pub tests: bool,
     pub backtrace: bool,
     pub code: String,
+}
+
+impl EditionRequest for ExecuteRequest {
+    fn edition(&self) -> Option<Edition> { self.edition }
 }
 
 #[derive(Debug, Clone)]
@@ -698,6 +720,10 @@ pub struct ClippyResponse {
 pub struct MiriRequest {
     pub code: String,
     pub edition: Option<Edition>,
+}
+
+impl EditionRequest for MiriRequest {
+    fn edition(&self) -> Option<Edition> { self.edition }
 }
 
 #[derive(Debug, Clone)]

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -853,14 +853,13 @@ mod test {
         assert!(resp.stdout.contains("nightly"));
     }
 
+    // Code that will only work in Rust 2018
     const EDITION_CODE: &str = r#"
-    mod foo {
-        pub fn bar() {}
+    macro_rules! foo {
+        ($($a:ident)?) => {}
     }
 
-    fn main() {
-        crate::foo::bar();
-    }
+    fn main() {}
     "#;
 
     #[test]
@@ -874,7 +873,7 @@ mod test {
         let sb = Sandbox::new()?;
         let resp = sb.execute(&req)?;
 
-        assert!(resp.stderr.contains("`crate` in paths is experimental"));
+        assert!(resp.stderr.contains("`?` is not a macro repetition operator"));
         Ok(())
     }
 
@@ -890,7 +889,7 @@ mod test {
         let sb = Sandbox::new()?;
         let resp = sb.execute(&req)?;
 
-        assert!(resp.stderr.contains("`crate` in paths is experimental"));
+        assert!(resp.stderr.contains("`?` is not a macro repetition operator"));
         Ok(())
     }
 
@@ -996,7 +995,7 @@ mod test {
         let resp = sb.compile(&req).expect("Unable to compile code");
 
         assert!(resp.code.contains("core::fmt::Arguments::new_v1"));
-        assert!(resp.code.contains("std::io::stdio::_print@PLT"));
+        assert!(resp.code.contains("std::io::stdio::_print@GOTPCREL"));
     }
 
     #[test]
@@ -1046,8 +1045,8 @@ mod test {
         let sb = Sandbox::new().expect("Unable to create sandbox");
         let resp = sb.clippy(&req).expect("Unable to lint code");
 
-        assert!(resp.stderr.contains("deny(eq_op)"));
-        assert!(resp.stderr.contains("warn(zero_divided_by_zero)"));
+        assert!(resp.stderr.contains("deny(clippy::eq_op)"));
+        assert!(resp.stderr.contains("warn(clippy::zero_divided_by_zero)"));
     }
 
     #[test]
@@ -1067,7 +1066,7 @@ mod test {
         let sb = Sandbox::new()?;
         let resp = sb.miri(&req)?;
 
-        assert!(resp.stderr.contains("pointer computed at offset 1"));
+        assert!(resp.stderr.contains("Pointer must be in-bounds and live at offset 1"));
         assert!(resp.stderr.contains("outside bounds of allocation"));
         assert!(resp.stderr.contains("which has size 0"));
         Ok(())

